compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2256,15 +2270,49 @@ impl<'a> Parser<'a> {
 
     /// Parses an `else { ... }` expression (`else` token already eaten).
     fn parse_else_expr(&mut self) -> PResult<'a, P<Expr>> {
-        let ctx_span = self.prev_token.span; // `else`
+        let else_span = self.prev_token.span; // `else`
         let attrs = self.parse_outer_attributes()?.take_for_recovery(); // For recovery.
         let expr = if self.eat_keyword(kw::If) {
             self.parse_if_expr(AttrVec::new())?
+        } else if self.check(&TokenKind::OpenDelim(Delimiter::Brace)) {
+            self.parse_simple_block()?
         } else {
-            let blk = self.parse_block()?;
-            self.mk_expr(blk.span, ExprKind::Block(blk, None), AttrVec::new())
+            let snapshot = self.clone();
```

Comment:
There's a [`create_snapshot_for_diagnostic`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_parse/parser/struct.Parser.html#method.create_snapshot_for_diagnostic) helper function for doing this that clears some of the internal state of the snapshot that might make it emit diagnostics when dropped.